lightning/src/ln/onchain_utils.rs

```diff
@@ -88,3 +123,649 @@ pub(crate) fn get_witnesses_weight(inputs: &[InputDescriptors]) -> usize {
+/// Preimage is only included as part of the witness in former case.
+#[derive(Clone, PartialEq)]
+pub(crate) struct HolderHTLCOutput {
+	preimage: Option<PaymentPreimage>,
+	amount: u64,
+}
+
+impl HolderHTLCOutput {
+	pub(crate) fn build(preimage: Option<PaymentPreimage>, amount: u64) -> Self {
+		HolderHTLCOutput {
+			preimage,
+			amount
+		}
+	}
+}
+
+impl Writeable for HolderHTLCOutput {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		self.preimage.write(writer)?;
+		writer.write_all(&byte_utils::be64_to_array(self.amount))?;
+		Ok(())
+	}
+}
+
+impl Readable for HolderHTLCOutput {
+	fn read<R: ::std::io::Read>(reader: &mut R) -> Result<Self, DecodeError> {
+		let preimage = Readable::read(reader)?;
+		let amount = Readable::read(reader)?;
+		Ok(HolderHTLCOutput {
+			preimage,
+			amount,
+		})
+	}
+}
+
+/// A struct to describe the channel output on the funding transaction.
+///
+/// witnessScript is used as part of the witness redeeming the funding utxo.
+#[derive(Clone, PartialEq)]
+pub(crate) struct HolderFundingOutput {
+	funding_redeemscript: Script,
+}
+
+impl HolderFundingOutput {
+	pub(crate) fn build(funding_redeemscript: Script) -> Self {
+		HolderFundingOutput {
+			funding_redeemscript,
+		}
+	}
+}
+
+impl Writeable for HolderFundingOutput {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		self.funding_redeemscript.write(writer)?;
+		Ok(())
+	}
+}
+
+impl Readable for HolderFundingOutput {
+	fn read<R: ::std::io::Read>(reader: &mut R) -> Result<Self, DecodeError> {
+		Ok(HolderFundingOutput {
+			funding_redeemscript: Readable::read(reader)?,
+		})
+	}
+}
+
+/// A wrapper encapsulating all in-protocol differing outputs categories.
+///
+/// The generic API offers access to outputs attributes (amount, weight, category) or allow
+/// transformation such as finalizing an input claiming the output.
+#[derive(Clone, PartialEq)]
+pub(crate) enum PackageSolvingData {
+	RevokedOutput(RevokedOutput),
+	CounterpartyHTLCOutput(CounterpartyHTLCOutput),
+	HolderHTLCOutput(HolderHTLCOutput),
+	HolderFundingOutput(HolderFundingOutput),
+}
+
+impl PackageSolvingData {
+	fn amount(&self) -> u64 {
+		let amt = match self {
+			PackageSolvingData::RevokedOutput(ref outp) => { outp.amount },
+			PackageSolvingData::CounterpartyHTLCOutput(ref outp) => { outp.htlc.amount_msat / 1000 },
+			PackageSolvingData::HolderHTLCOutput(..) => { 0 },
+			PackageSolvingData::HolderFundingOutput(..) => { 0 },
+		};
+		amt
+	}
+	fn weight(&self) -> usize {
+		let weight = match self {
+			PackageSolvingData::RevokedOutput(ref outp) => { get_witnesses_weight(&[outp.input_descriptor]) },
+			PackageSolvingData::CounterpartyHTLCOutput(ref outp) => { get_witnesses_weight(if outp.preimage.is_some() { &[InputDescriptors::OfferedHTLC] } else { &[InputDescriptors::ReceivedHTLC]}) },
+			// Note: Currently, weights of holder outputs spending witnesses aren't used
```

Comment:
If this is unused can we at least have a `debug_assert!(false)` instead of returning a bogus value?

Reply:
Amount is exercised in `generate_claim_tx` so added a debug_assert(amt == 0) otherwise added for weight.